### PR TITLE
Don't remove all mirror repository's releases when mirroring

### LINKED ISSUE #22066: Releases are deleted when mirror repository is synchronized
### Description

By some reason the created releases are deleted when mirror repository is synchronized even when the tags have not been recreated.

#### Steps

1. Sign-in to Gitea
2. Click "**+**"
3. Click "**New Migration**"
4. Click "**GitHub**"
5. Put `https://github.com/tailix/libkernaux` in the field "**Migrate / Clone From URL**"
6. Check "**This repository will be a mirror**"
7. Click the button "**Create repository**"
8. Click "**7 tags**" in the bar (the number may be different)
9. Click "**New release**" on tag "**v0.6.1**"
10. Put `foobar` in the field "**Title**"
11. Click "**Publish release**"
12. Click on the tab "**Settings**"
13. Click the button "**Synchronize Now**"
14. Wait a few (2-5) minutes
15. Refresh the page

#### Expected

You should see the label "**Releases [1]**" (the number as a badge) among the tabs.

#### Got

You see the label "**Releases**" (without a number) among the tabs.

---

### Gitea Version

1.17.3 (self-hosted), 1.19.0+dev-166-g0a85537c7 (try.gitea.io)

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

2.37.2

### Operating System

Ubuntu 22.10

### How are you running Gitea?

https://try.gitea.io & self-hosted

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by lunny**
Fix #22066

# Purpose

This PR fix the releases will be deleted when mirror repository sync the tags. 

# The problem

In the previous implementation of #19125. All releases record in databases of one mirror repository will be deleted before sync.
Ref: https://github.com/go-gitea/gitea/pull/19125/files#diff-2aa04998a791c30e5a02b49a97c07fcd93d50e8b31640ce2ddb1afeebf605d02R481

# The Pros

This PR introduced a new method which will load all releases from databases and all tags on git data into memory. And detect which tags needs to be inserted, which tags need to be updated or deleted. Only tags releases(IsTag=true) which are not included in git data will be deleted, only tags which sha1 changed will be updated. So it will not delete any real releases include drafts.

# The Cons

The drawback is the memory usage will be higher than before if there are many tags on this repository. This PR defined a special release struct to reduce columns loaded from database to memory.

### CODE REVIEW

**Review by wxiaoguang — COMMENTED** [review]
(outdated)

<details>

~~Although I was requested for reviewing, but: no test, no comment, no description, so no idea how to review.~~

</details>

**Comment by lunny** [comment]
> Although I was requested for reviewing, but: no test, no comment, no description, so no idea how to review.

Sorry. I have reverted unnecessary refactor changes and also added a unit test. Please review it again.

**Review by wxiaoguang — APPROVED** [review]

File: modules/repository/repo_test.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		{
+			ID:      4,
+			TagName: "v1.1.0",
+			Sha1:    "53ab18dcecf4152b58328d1f47429510eb414d50",
+		},
+	}
+
+	inserts, deletes, updates := calcSync(gitTags, dbReleases)
+	assert.EqualValues(t, 1, len(inserts), "inserts")
+	assert.EqualValues(t, *gitTags[2], *inserts[0], "inserts equal")
```
Comment (6543) [inline modules/repository/repo_test.go]:
```suggestion
	if assert.EqualLen(t, inserts, 1) {
		assert.EqualValues(t, *gitTags[2], *inserts[0], "inserts equal")
	}
```

File: modules/repository/repo_test.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+	assert.EqualValues(t, 1, len(deletes), "deletes")
```
Comment (6543) [inline modules/repository/repo_test.go]: same ...

File: modules/repository/repo_test.go (hunk 1 above)
Comment (wolfogre) [reply modules/repository/repo_test.go]: I don't think we have to do this since it's just testing. If it panics, just let it panic.

**Review by wolfogre — APPROVED** [review]

**Pull request merged**